src/lib.rs

```diff
@@ -70,12 +71,42 @@ fn shrink<T>(slice: &mut &mut [T], len: usize) {
 fn is_token(b: u8) -> bool {
     b > 0x1F && b < 0x7F
 }
+
 macro_rules! byte_map {
     ($($flag:expr,)*) => ([
         $($flag != 0,)*
     ])
 }
 
+// ASCII codes to accept URI string.
+// i.e. A-Z a-z 0-9 !#$%&'*+-._();:@=,/?[]~
+// TODO: Make a stricter checking for URI string?
+static URI_MAP: [bool; 256] = byte_map![
```

Comment:
Just cause I'm curious, why is this different from just parsing as a `token`?

Reply:
From [RFC7230 Section 3.2.6](https://tools.ietf.org/html/rfc7230#section-3.2.6), we can get the formalize description of token:

```
token  = 1*tchar

tchar  = "!" / "#" / "$" / "%" / "&" / "'" / "*"
            / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~"
            / DIGIT / ALPHA
            ; any VCHAR, except delimiters
```

Current `is_token(b: u8) -> bool;` is not satisfy this def, I think that should be changed. 

For URI matching, depends on the safeguard requirement, the accepted charset may differ, IMO, we should be careful here, maybe we can provide a `strict mode`, which is an optional feature here?